wireguard/wireguard.go

```diff
@@ -242,119 +272,201 @@ func (w *Wireguard) EndpointUpdate(name string, ipv4Addr ip.Addr) {
-		w.logCxt.Debug("Node CIDR added which is not programmed")
-		update.allowedCidrsAdded.Add(cidr)
-		w.cidrToNodeNameUpdates[cidr] = name
+		w.peerAllowedCIDRAdd(name, cidr)
 	}
-	w.setPeerUpdate(name, update)
 }
 
-func (w *Wireguard) EndpointAllowedCIDRRemove(cidr ip.CIDR) {
-	w.logCxt.Debugf("EndpointAllowedCIDRRemove: cidr=%v", cidr)
+func (w *Wireguard) RouteRemove(cidr ip.CIDR) {
+	logCxt := log.WithField("cidr", cidr)
+	logCxt.Debug("RouteRemove")
 	if !w.config.Enabled {
-		w.logCxt.Debug("Not enabled - ignoring")
+		log.Debug("Not enabled - ignoring")
 		return
 	}
 
 	// Determine which node this CIDR belongs to. Check the updates first and then the processed.
-	name, ok := w.cidrToNodeNameUpdates[cidr]
+	name, ok := w.cidrToNodeName[cidr]
 	if !ok {
-		w.logCxt.Debugf("CIDR not found as node update, checking current configuration")
-		name, ok = w.cidrToNodeName[cidr]
-		if !ok {
-			// The wireguard manager filters out some of the CIDR updates, but not the removes, so it's possible to get
-			// CIDR removes for which we have seen no corresponding add.
-			w.logCxt.Debugf("CIDR remove update but not associated with a node: %v", cidr)
-			return
+		// The wireguard manager filters out some of the CIDR updates, but not the removes, so it's possible to get
+		// CIDR removes for which we have seen no corresponding add.
+		logCxt.Debug("CIDR remove update but not associated with a node")
+		return
+	}
+	logCxt.WithField("node", name).Debug("CIDR associated with node")
+	w.routeRemove(name, cidr)
+}
+
+func (w *Wireguard) routeRemove(name string, cidr ip.CIDR) {
+	// Remove the CIDR->node mapping.
+	delete(w.cidrToNodeName, cidr)
+
+	// If this is the local node then remove as a local workload CIDR, otherwise remove as a peer CIDR.
+	if name == w.hostname {
+		w.localWorkloadCIDRRemove(cidr)
+	} else {
+		w.peerAllowedCIDRRemove(name, cidr)
+	}
+}
+
+// Add a local workload CIDR. These CIDRs are used for the source-matched wireguard routes to limit wireguard encryption
+// to traffic to/from local workloads. The workload CIDRs may overlap, in which case the updateAndApplyRouteRules
+// method will determine the minimal set of non-overlapping CIDRs.
+func (w *Wireguard) localWorkloadCIDRAdd(cidr ip.CIDR) {
+	log.WithField("cidr", cidr).Debug("localWorkloadCIDRAdd")
+	// Split the local CIDRs into actual /32 workload IPs and the CIDR blocks for the node. We assume the CIDR blocks
+	// are not overlapping, and so we add rules for each CIDR to route to wireguard, and only include the /32 workload
+	// IPs if not covered by the CIDR blocks.
+	if cidr.Prefix() == ipPrefixLen {
+		w.localIPs.Add(cidr.Addr())
+	} else {
+		w.localCIDRs.Add(cidr)
+	}
+	// Only flag the CIDRs for update if it not wholly covered by the already filtered local CIDRs.
+	contained := false
+	w.localCIDRsFiltered.Iter(func(item interface{}) error {
+		filtered := item.(ip.CIDR).ToIPNet()
+		if filtered.Contains(cidr.ToIPNet().IP) {
```

Comment:
By this test, I think `10.0.0.0/32` would contain `10.0.0.0/16`, which is wrong.  You need to check the length too.

Reply:
heh.  oh yes, good spot!